lib/px/utils/upload.lua

```diff
@@ -0,0 +1,334 @@
+-- Copyright (C) Yichun Zhang (agentzh)
+
+local req_socket = ngx.req.socket
+local match = string.match
+local setmetatable = setmetatable
+local type = type
+local ngx_var = ngx.var
+local ngx_init_body = ngx.req.init_body
+local ngx_finish_body = ngx.req.finish_body
+local ngx_append_body = ngx.req.append_body
+
+local _M = { _VERSION = '0.10' }
+
+
+local CHUNK_SIZE = 4096
+local MAX_LINE_SIZE = 512
+
+local STATE_BEGIN = 1
+local STATE_READING_HEADER = 2
+local STATE_READING_BODY = 3
+local STATE_EOF = 4
+
+
+local mt = { __index = _M }
+
+local state_handlers
+
+
+local function get_boundary()
+    local header = ngx_var.content_type
+    if not header then
+        return nil
+    end
+
+    if type(header) == "table" then
+        header = header[1]
+    end
+
+    local m = match(header, ";%s*boundary=\"([^\"]+)\"")
+    if m then
+        return m
+    end
+
+    return match(header, ";%s*boundary=([^\",;]+)")
+end
+
+
+function _M.new(self, chunk_size, max_line_size, restore_body_buffer)
```

Comment:
I tried to understand why calling it new with no success. It creates an object, as most functions do. Why new?

Reply:
`_m.new` is Lua constructor for `_M` object

Reply:
So why there is so much logic inside a constructor 🤔